### ISSUE TITLE
Allow registration of custom attributes onto existing entities.

### ISSUE THREAD

**Issue by i509VCB** [FabricMC/fabric#626]
This was an API gap I discovered while experimenting with custom entity attributes.

So lets assume we create a new attribute, piercing for example.

So we register it with below code, not much of a challenge to create an entity attribute.
```java
public static final EntityAttribute PIERCING = register("generic.piercing", 5.0D);

private static EntityAttribute register(String path, double maxValue) {
...
```

So per 1.16 snapshots, entities must now have the attribute registered through the `DefaultAttributeRegistry.Builder`. One issue is lets assume I wanted to register the piercing attribute to all entities. So there are two solutions:

A. Use Mixin to add the attribute to all entities. This requires in some cases with modpacks hundreds of exhaustive mixins.

B. Replace every entity's default attribute container with a new one that contains our own attribute as well. This is also exhaustive and not very compatible with other mods who may also want to add their own attributes to entities.

So the solution I see is to allow adding custom entity attributes to the builder. The specifics of this can be a challenge, such as whether we should allow this to be done by entity type, or entity class. Below is two ideas I did think of, but more ideas are welcome

```java
// EntityType based
CreateDefaultAttributesCallback.event(EntityType.CAT).register(builder -> {
    builder.add(CustomAttributes.PIERCING, 3.0D);
});

// Entity class based
CreateDefaultAttributesCallback.event(HostileEntity.class).register(builder -> {
    builder.add(CustomAttributes.BLUDGEONING).add(CustomAttributes.PIERCING);
});
```

**Comment by liach** [comment]
For classes, for now you can probably mixin to like the static `CatEntity.createCatAttributes` etc and call `add` on the return value of the method, etc.

Should you do it by entity type, it should be fine too, but it is a bit hard to convert existing default attribute container to a builder.

**Comment by i509VCB** [comment]
Yeah that is a bit of a challenge. The big thing is that the entity attributes are initialized during bootstrap, which is before the server entrypoint but after the client entrypoint. And a single mod directly accessing the DefaultAttributeRegistry would load everything before other mods had the chance

**Comment by NaymDev** [comment]
Is there a solution to this?

**Comment by aniomis** [comment]
You can add an attribute to all LivingEntity by injecting into the method `LivingEntity.createLivingAttributes()`, which builds the base attributes. It seems to work fine, but I'm not sure whether it might cause conflicts. It looks something like this:

```java
@Mixin(LivingEntity.class)
public class SpecialAttributeForAllLivingEntity {
    @Inject(at = @At("TAIL"), method = "createLivingAttributes")
    private static void createLivingAttributes(CallbackInfoReturnable<DefaultAttributeContainer.Builder> cir) {
        cir.getReturnValue().add(EntityAttributes.MY_SPECIAL_ATTRIBUTE, 0.0);
    }
}
```

(Updated the code based on TheDeathlyCow's comment below.)

**Comment by TheDeathlyCow** [comment]
I can tell you from experience that this will cause conflicts. The setReturnValue call will cancel the method for any later mods adding a mixin point here to do something similar. Either use @ModifyReturnValue or omit the cancelling call. The builder is mutable so the change will still be applied even if you don't use setReturnValue

**Comment by aniomis** [comment]
Thank you so much for the feedback! I didn't realize the builder was just mutating itself. I updated the code in my comment so that nobody uses the conflict-inducing code.

Again I really appreciate the help. :)
